File: internal/filter/filter_test.go
package filter

import (
	"regexp"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestApply(t *testing.T) {
	dirs := []string{"src/app", "src/lib", "docs", "tools/cli"}

	t.Run("includeOnly filters to matching dirs", func(t *testing.T) {
		result := Apply(dirs, Options{IncludeOnly: []string{"src/app", "docs"}})
		assert.Equal(t, []string{"src/app", "docs"}, result)
	})

	t.Run("includeOnly matches basename", func(t *testing.T) {
		result := Apply(dirs, Options{IncludeOnly: []string{"app"}})
		assert.Equal(t, []string{"src/app"}, result)
	})

	t.Run("excludeOnly removes matching dirs", func(t *testing.T) {
		result := Apply(dirs, Options{ExcludeOnly: []string{"docs"}})
		assert.Equal(t, []string{"src/app", "src/lib", "tools/cli"}, result)
	})

	t.Run("excludeOnly matches basename", func(t *testing.T) {
		result := Apply(dirs, Options{ExcludeOnly: []string{"cli"}})
		assert.Equal(t, []string{"src/app", "src/lib", "docs"}, result)
	})

	t.Run("includePattern filters by regex", func(t *testing.T) {
		result := Apply(dirs, Options{IncludePattern: regexp.MustCompile(`^src/`)})
		assert.Equal(t, []string{"src/app", "src/lib"}, result)
	})

	t.Run("excludePattern removes by regex", func(t *testing.T) {
		result := Apply(dirs, Options{ExcludePattern: regexp.MustCompile(`^docs`)})
		assert.Equal(t, []string{"src/app", "src/lib", "tools/cli"}, result)
	})

	t.Run("combined filters apply sequentially", func(t *testing.T) {
		result := Apply(dirs, Options{
			ExcludeOnly:    []string{"docs"},
			IncludePattern: regexp.MustCompile(`^src/`),
		})
		assert.Equal(t, []string{"src/app", "src/lib"}, result)
	})

	t.Run("empty options returns all dirs", func(t *testing.T) {
		result := Apply(dirs, Options{})
		assert.Equal(t, dirs, result)
	})

	t.Run("does not mutate input", func(t *testing.T) {
		original := []string{"a", "b", "c"}
		input := make([]string, len(original))
		copy(input, original)
		Apply(input, Options{IncludeOnly: []string{"a"}})
		assert.Equal(t, original, input)
	})
}

func TestParseFilterList(t *testing.T) {
	t.Run("splits comma-separated values", func(t *testing.T) {
		assert.Equal(t, []string{"a", "b", "c"}, ParseFilterList("a,b,c"))
	})

	t.Run("trims whitespace", func(t *testing.T) {
		assert.Equal(t, []string{"a", "b"}, ParseFilterList(" a , b "))
	})

	t.Run("filters empty strings", func(t *testing.T) {
		assert.Equal(t, []string{"a", "b"}, ParseFilterList("a,,b,"))
	})

	t.Run("returns nil for empty input", func(t *testing.T) {
		assert.Nil(t, ParseFilterList(""))
	})
}

func TestParseFilterPattern(t *testing.T) {
	t.Run("compiles valid regex", func(t *testing.T) {
		re, err := ParseFilterPattern(`^src/`)
		require.NoError(t, err)
		assert.True(t, re.MatchString("src/app"))
	})

	t.Run("returns nil for empty input", func(t *testing.T) {
		re, err := ParseFilterPattern("")
		require.NoError(t, err)
		assert.Nil(t, re)
	})

	t.Run("returns error for invalid regex", func(t *testing.T) {
		_, err := ParseFilterPattern("[invalid")
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "invalid regex pattern")
	})
}

func TestCreateFilterOptions(t *testing.T) {
	t.Run("creates options from strings", func(t *testing.T) {
		opts, err := CreateFilterOptions("a,b", "c", `^src/`, `test$`)
		require.NoError(t, err)
		assert.Equal(t, []string{"a", "b"}, opts.IncludeOnly)
		assert.Equal(t, []string{"c"}, opts.ExcludeOnly)
		assert.NotNil(t, opts.IncludePattern)
		assert.NotNil(t, opts.ExcludePattern)
	})

	t.Run("handles empty strings", func(t *testing.T) {
		opts, err := CreateFilterOptions("", "", "", "")
		require.NoError(t, err)
		assert.Nil(t, opts.IncludeOnly)
		assert.Nil(t, opts.ExcludeOnly)
		assert.Nil(t, opts.IncludePattern)
		assert.Nil(t, opts.ExcludePattern)
	})
}

func TestFilterFromLoopRc(t *testing.T) {
	dirs := []string{"src/app", "src/lib", "docs", "examples"}

	t.Run("filters matching dirs", func(t *testing.T) {
		result := FilterFromLoopRc(dirs, []string{"docs", "examples"})
		assert.Equal(t, []string{"src/app", "src/lib"}, result)
	})

	t.Run("returns all dirs when ignore is empty", func(t *testing.T) {
		result := FilterFromLoopRc(dirs, []string{})
		assert.Equal(t, dirs, result)
	})

	t.Run("matches basename", func(t *testing.T) {
		result := FilterFromLoopRc(dirs, []string{"app"})
		assert.Equal(t, []string{"src/lib", "docs", "examples"}, result)
	})
}

File: tests/unit/core/filter.test.ts
import { describe, it, expect } from 'vitest';
import {
  applyFilters,
  parseFilterList,
  parseFilterPattern,
  createFilterOptions,
} from '../../../src/core/filter.js';

describe('filter', () => {
  describe('applyFilters', () => {
    const directories = ['api', 'web', 'shared', 'docs', 'test-utils'];

    it('returns all directories when no filters applied', () => {
      const result = applyFilters(directories, {});
      expect(result).toEqual(directories);
    });

    describe('includeOnly', () => {
      it('filters to only specified directories', () => {
        const result = applyFilters(directories, {
          includeOnly: ['api', 'web'],
        });
        expect(result).toEqual(['api', 'web']);
      });

      it('handles non-existent directories gracefully', () => {
        const result = applyFilters(directories, {
          includeOnly: ['api', 'nonexistent'],
        });
        expect(result).toEqual(['api']);
      });

      it('matches on basename for nested paths', () => {
        const dirs = ['libs/api', 'libs/web'];
        const result = applyFilters(dirs, {
          includeOnly: ['api'],
        });
        expect(result).toEqual(['libs/api']);
      });
    });

    describe('excludeOnly', () => {
      it('removes specified directories', () => {
        const result = applyFilters(directories, {
          excludeOnly: ['docs', 'test-utils'],
        });
        expect(result).toEqual(['api', 'web', 'shared']);
      });

      it('handles non-existent directories gracefully', () => {
        const result = applyFilters(directories, {
          excludeOnly: ['nonexistent'],
        });
        expect(result).toEqual(directories);
      });
    });

    describe('includePattern', () => {
      it('includes directories matching regex', () => {
        const result = applyFilters(directories, {
          includePattern: /^(api|web)$/,
        });
        expect(result).toEqual(['api', 'web']);
      });

      it('works with partial matches', () => {
        const result = applyFilters(directories, {
          includePattern: /test/,
        });
        expect(result).toEqual(['test-utils']);
      });
    });

    describe('excludePattern', () => {
      it('excludes directories matching regex', () => {
        const result = applyFilters(directories, {
          excludePattern: /^test-/,
        });
        expect(result).toEqual(['api', 'web', 'shared', 'docs']);
      });

      it('can exclude multiple matches', () => {
        const result = applyFilters(directories, {
          excludePattern: /^(docs|shared)$/,
        });
        expect(result).toEqual(['api', 'web', 'test-utils']);
      });
    });

    describe('combined filters', () => {
      it('applies includeOnly before excludeOnly', () => {
        const result = applyFilters(directories, {
          includeOnly: ['api', 'web', 'shared'],
          excludeOnly: ['shared'],
        });
        expect(result).toEqual(['api', 'web']);
      });

      it('applies all filters in order', () => {
        const result = applyFilters(directories, {
          includeOnly: ['api', 'web', 'shared', 'test-utils'],
          excludeOnly: ['shared'],
          excludePattern: /test/,
        });
        expect(result).toEqual(['api', 'web']);
      });
    });
  });

  describe('parseFilterList', () => {
    it('parses comma-separated list', () => {
      const result = parseFilterList('api,web,shared');
      expect(result).toEqual(['api', 'web', 'shared']);
    });

    it('trims whitespace', () => {
      const result = parseFilterList(' api , web , shared ');
      expect(result).toEqual(['api', 'web', 'shared']);
    });

    it('filters empty strings', () => {
      const result = parseFilterList('api,,web');
      expect(result).toEqual(['api', 'web']);
    });

    it('returns undefined for empty input', () => {
      expect(parseFilterList('')).toBeUndefined();
      expect(parseFilterList(undefined)).toBeUndefined();
    });
  });

  describe('parseFilterPattern', () => {
    it('creates regex from string', () => {
      const result = parseFilterPattern('^test');
      expect(result).toBeInstanceOf(RegExp);
      expect(result?.test('test-utils')).toBe(true);
    });

    it('returns undefined for empty input', () => {
      expect(parseFilterPattern('')).toBeUndefined();
      expect(parseFilterPattern(undefined)).toBeUndefined();
    });

    it('throws for invalid regex', () => {
      expect(() => parseFilterPattern('[')).toThrow('Invalid regex pattern');
    });
  });

  describe('createFilterOptions', () => {
    it('creates filter options from string inputs', () => {
      const result = createFilterOptions({
        includeOnly: 'api,web',
        excludeOnly: 'docs',
        includePattern: '^lib',
        excludePattern: 'test',
      });

      expect(result.includeOnly).toEqual(['api', 'web']);
      expect(result.excludeOnly).toEqual(['docs']);
      expect(result.includePattern).toBeInstanceOf(RegExp);
      expect(result.excludePattern).toBeInstanceOf(RegExp);
    });

    it('handles undefined inputs', () => {
      const result = createFilterOptions({});

      expect(result.includeOnly).toBeUndefined();
      expect(result.excludeOnly).toBeUndefined();
      expect(result.includePattern).toBeUndefined();
      expect(result.excludePattern).toBeUndefined();
    });
  });
});
